### ISSUE TITLE
okta-react single log out not working

### ISSUE THREAD

**Issue by PaulStat** [okta/okta-oidc-js#397]
This issue was spotted in the okta-react package, I believe it to be a bug.

Using these libraries if for example a user signs in to a React application via Okta then in a separate browser tab browses to the Okta dashboard and clicks sign out. The user can still browse the React application.

However if the user has implemented the logout functionality of Auth.js then both the Okta session and tokens are cleared.

I would expect the killing of the Okta session also to clear any tokens. I'd like to suggest the following amendment to the isAuthenticated function of Auth.js

` async isAuthenticated() {
    const sessionExists = await this._oktaAuth.session.exists();

    // If there could be tokens in the url
    if (window.location && window.location.hash && containsAuthTokens.test(window.location.hash)) return null;

    const hasTokens = !!(await this.getAccessToken()) || !!(await this.getIdToken());

    if (hasTokens && !sessionExists) {
      this._oktaAuth.tokenManager.clear();
      this._history.push('/');
      return false;
    }

    return hasTokens;
  }`

**Comment by PaulStat** [comment]
I have also noticed that this library doesn't seem to utilise SSO? Am I missing something here? Eg. log in to Okta dashboard, open react SPA using these libraries and we have to reauthenticate? Again this would be simple to add to the above example

```
async isAuthenticated() {
    // If there could be tokens in the url
    if (window.location && window.location.hash && containsAuthTokens.test(window.location.hash)) return null;

    const sessionExists = await this._oktaAuth.session.exists();

    const hasTokens = !!(await this.getAccessToken()) || !!(await this.getIdToken());
    
    //SSO    
    if (sessionExists && !hasTokens) {
      this._oktaAuth..token.getWithRedirect(
          {
              responseType: ['id_token', 'token'],
              scope: ['profile', 'email'],
          }
      );
    }

    // SLO
    if (hasTokens && !sessionExists) {
      this._oktaAuth.tokenManager.clear();
      this._history.push('/');
      return false;
    }

    return hasTokens;
  }
```

**Comment by swiftone** [comment]
Thanks for the detailed comments - Reviewing your suggestions is on our TODO pile, but I wanted to make a quick mention here so that it doesn't look like you are being ignored while we are busy dealing with other tasks.

The logout functionality isn't a bug, per se, but you are expressing a common desire.  We've already implemented okta logout (SLO as you express it here) in our oidc-middleware for express and will be bringing it to our other SDKs.

**Comment by sibelius** [comment]
@swiftone so the recommended way to logout user is to do it on backend right?

**Comment by swiftone** [comment]
@sibelius - That depends on what you mean by "logout".  Hopefully this wall of text has some useful bits for you, but the answer to "how do I revoke the tokens" is the last part.

* To have your app check against Okta again, just have your application forget any tokens.  Many people can find this confusing, though, as when they attempt to re-login to the application they will redirect to Okta, and if they have a valid active session with Okta they will redirect back to the application without ever being prompted for a user/password, so they may not even notice that Okta was checked.

* To require re-authentication (i.e. to require they use their username/password) you have to tell Okta to remove their active session.   This does NOT revoke any existing tokens, and applications that have active tokens will continue to consider the user "logged in".  This "still considered logged in" is because avoiding network checks on every call is part of the concept of tokens - if that doesn't match your use-case, check further down.
  * On the front end requiring re-authentication can be done by 
     * (What the okta-react package currently does) Sending a web service call to DELETE /sessions/me.  See https://developer.okta.com/docs/api/resources/sessions/#close-current-session to see the requirements, in particular the Authorization header.  (Note: there is a current bug that throws an error if you do this when the user is already logged out that I'm in the process of fixing)
     * Alternatively, you could redirect them to the Okta /logout endpoint: https://developer.okta.com/docs/api/resources/oidc/#logout
  * On the backend you can make the same call to DELETE /sessions/me

* To render the tokens "invalid" you can tell Okta to revoke it ( see 
 https://developer.okta.com/docs/api/resources/oidc/#revoke, frontend or backend service calls ) BUT an application with a token won't be aware this has happened unless it checks.  (Again, the idea is that a token can be used and validated without repeated network calls).  You can check that a token is valid by asking Okta if the token remains valid ( https://developer.okta.com/docs/api/resources/oidc/#introspect ) This can be done frontend or backend with a service call.

TL;DR: To make a user that has logged out elsewhere be considered logged out by your app when they otherwise have an unexpired token, after a user logs out, the token has to be revoked AND the application using the token has to make a service call to Okta (after the revocation) to revalidate the token.  Because of the repeated network calls I don't recommend doing so as a regular practice and instead only doing so before higher security-threshold events (such as changing contact information), if at all.

**Comment by sibelius** [comment]
My use case is to require re-authentication with okta as well

Right now I'm calling logout like this in my React app

`auth.logout` (based on this code https://github.com/okta/okta-oidc-js/tree/master/packages/okta-react#show-login-and-logout-buttons)

this does not force `re-authentication`

So the right solution to me is to send user to `/logout` endpoint by myself right?

Another question, as my application is SPA, I don't have a client secret to `revoke` tokens on backend
So how can I remote kill a logged user from okta? remote kill would be the backend removing okta session from an user that needs to login again

**Comment by swiftone** [comment]
> this does not force re-authentication

Sorry to be pedantic here, I need to make sure I have the details correct:

When the user logs off using the `auth.logout` call, for them to login in to the same app again SHOULD require re-entering their username/password.  (Not talking about any apps they are already logged into).  Are you saying this is not working for you?

Put another way:
* User logs into App A 
* User logs out of App A using the `auth.logout` call 
* User attempts to log in to App A (same browser tab) - they should be prompted for username/password

> So the right solution to me is to send user to /logout endpoint by myself right

That should not be required.  That's an alternate method to do what should already be happening.

> as my application is SPA, I don't have a client secret to revoke tokens on backend

A SPA app can make the call to revoke without the Authorization header - Okta will see that the token is for a application clientId that doesn't have a secret and will use the token itself to validate that the call is legitimate.

**Comment by sibelius** [comment]
This is what is happening to me

- User logs into my application A
- redirect to Okta, and needs to enter email and password
- redirect back to application A
- User logout from my application A (I call auth.logout() here)
- User logs into my application A again
- redirect to Okta, does not need to enter email/password (I need the user to put the email and password again on Okta screen)
- redirect back to application A

**Comment by swiftone** [comment]
@sibelius Your expectations for what should be happening (versus what you are seeing) are correct - I'm investigating.

**Comment by swiftone** [comment]
@sibelius I've been unable to reproduce your error - Can you offer more details about login/logout calls?  Also, what versions of the okta libraries are you running?

**Comment by swiftone** [comment]
@sibelius - Sorry for the comment spam - could you report your issue as a separate issue, as it appears to be distinct from this issue (which is about token revocation)?  Thanks in advance!
